src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs

```diff
@@ -295,6 +295,33 @@ private async Task ForceAuthenticationAsync<TIOAdapter>(bool receiveFirst, byte[
             }
             try
             {
+#if TARGET_OSX
+                if (SslStreamPal.ShouldUseAsyncSecurityContext(_sslAuthenticationOptions))
+                {
+                    // For Network Framework, we need to select client certificate before handshake
+                    if (_sslAuthenticationOptions.IsClient)
+                    {
+                        X509Certificate2? selectedCert = SelectClientCertificate();
+                        if (selectedCert != null && _sslAuthenticationOptions.CertificateContext == null)
+                        {
+                            // Build certificate context if not already provided
+                            _sslAuthenticationOptions.CertificateContext = SslStreamCertificateContext.Create(selectedCert);
+                        }
+                    }
+
+                    Task<Exception?> handshakeTask = SslStreamPal.AsyncHandshakeAsync(ref _securityContext, _sslAuthenticationOptions, InnerStream, cancellationToken);
```

Comment:
I don't really like the duplication here .... but I can live with that if we must. I would much prefer to integrate with this function ins some platform safe matter - like nullable environment that gets ignored on all other platforms.

Reply:
I'm not sure I understood the duplication you are referring here.
You're asking about to re-use the codepath we already have in here with more additions instead of making NW use completely separate codepath?